server/services/v1/database/query_index.go

```diff
@@ -0,0 +1,294 @@
+
+			if err != nil {
+				log.Err(err).Msgf("Failed to index field name: %s", field.FieldName)
+				continue
+			}
+
+			rows = []FieldIndex{*row}
+		}
+
+		for _, row := range rows {
+			indexKey := newKeyWithId(id, coll.EncodedName, coll.Indexes.QueryIndex.Name, "kvs", row.name, coll.IdxVer, row.value.AsInterface(), row.pos)
+
+			indexedKeys = append(indexedKeys, indexKey)
+			rowCount += 1
+			sizeIncrease += int64(len(indexKey.SerializeToBytes()))
+		}
+	}
+
+	return
+}
+
+func indexField(doc []byte, fieldName string, dataType schema.FieldType, pos int, keyPath ...string) (*FieldIndex, error) {
+	val, _, _, err := jsonparser.Get(doc, keyPath...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	row, err := newFieldIndex(dataType, fieldName, val, pos)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return row, nil
+}
+
+func IndexNestedField(doc []byte, topField string, pos int) []FieldIndex {
+	var indexedFields []FieldIndex
+	processor := func(key []byte, value []byte, jsonDataType jsonparser.ValueType, offset int) error {
+		fieldType := schema.ToFieldType(jsonDataType.String(), "", "")
+
+		switch fieldType {
+		case schema.ArrayType:
+			// Do not index into nested arrays
+			return nil
+		case schema.ObjectType:
+			IndexNestedField(value, topField+"."+string(key), pos)
+		default:
+			fieldName := topField + "." + string(key)
+			row, err := newFieldIndex(fieldType, fieldName, value, pos)
+
+			if err != nil {
+				log.Err(err).Msgf("Failed to index field name: %s", fieldName)
+				return err
+			}
+
+			indexedFields = append(indexedFields, *row)
+		}
+
+		return nil
+	}
+
+	jsonparser.ObjectEach(doc, processor)
+
+	return indexedFields
+}
+
+func indexArray(doc []byte, field *schema.QueryableField, keyPath []string) []FieldIndex {
+	pos := 0
+	var rows []FieldIndex
+	processor := func(value []byte, jsonDataType jsonparser.ValueType, offset int, err error) {
+		fmt.Println("indexing", field.FieldName)
+		toFieldType := schema.ToFieldType(jsonDataType.String(), "", "")
+		if toFieldType == schema.ObjectType {
+			indexedFields := IndexNestedField(value, field.FieldName, pos)
+
+			rows = append(rows, indexedFields...)
+
+		} else {
+			indexedField, err := newFieldIndex(field.SubType, field.FieldName, value, pos)
+			if err != nil {
+				log.Err(err).Msgf("Failed to index field name: %s", field.FieldName)
+				return
+			}
+			rows = append(rows, *indexedField)
+		}
+
+		pos += 1
+	}
+
+	jsonparser.ArrayEach(doc, processor, keyPath...)
+	return rows
+}
+
+func collateString(rawValue []byte) []byte {
+	collator := collate.New(language.English)
+
+	// Only index up to MAX_STRING_LEN of the string
+	if len(rawValue) > MAX_STRING_LEN {
+		rawValue = rawValue[:MAX_STRING_LEN+1]
+	}
+
+	var buf collate.Buffer
+	collated := collator.Key(&buf, rawValue)
+
+	return collated
+}
+
+func newKeyWithId(id []interface{}, table []byte, indexParts ...interface{}) keys.Key {
+	indexParts = append(indexParts, id...)
```

Comment:
It would be better to add a comment on this method that the id is added at the end which is a primary key. Better to refer to it as a primary key.